CS4850Group5A/VirtuDoc
# Outbound email abstraction layer

# Summary
- Added mock SMTP server to local dev stack for email testing (Available on port 3000).
- Added MailService class for one-line email sending.
- Added EmailDTO class for simplified message construction.
- Added TemplateEngine bean for HTML rendering outside of the normal MVC flow.
- Added configuration settings for MailGun integration in production.

Closes #15

# How To Test
- Verify that all JUnit tests pass in a bot comment below.
- Due to a bug that prevents Mockito from capturing certain legacy `com.sun.*` method invocations, the JUnit tests do not cover the `MailService` class when you run the JUnit tests natively (i.e. no Docker Compose). You can test the email sending components manually through the following steps.
1. Add the following to `src/main/java/com.virtudoc.web/controller/SampleEntityController.java`:
```java
@Autowired
private MailService mailService;

@GetMapping(path="/mail")
public @ResponseBody String testMailService() {
    EmailDTO email = new EmailDTO("test@example.com", "test subject", "mail/test");
    try {
        mailService.SendEmail(email);
    } catch (Exception e) {
        return "failed";
    }
    return "sent";
}
```
2. Launch the Docker Compose stack, including the `email` element.
3. Visit `localhost:8080/sampleentity/mail` and confirm that you see the word `sent`.
4. Open the mock SMTP server at `localhost:3000` and confirm that you see a test email in your inbox.

# Additional Comments
I know the original plan was SendGrid, but we now have to use MailGun. I accidentally deleted the SendGrid addon in the Heroku dashboard, and due to a bug in the SendGrid API, our Heroku app is now soft-banned from using SendGrid. Fortunately, MailGun is also free and offers more or less the same features while offering native integration with Heroku. My bad.
